src/hooks/useAddressGeneration.tsx

```diff
@@ -0,0 +1,194 @@
+        ...address,
+        isDefault: false,
+        label: labelPrefix ? `${labelPrefix} ${address.group}` : '',
+        color: labelColor ?? randomLabelColor
+      }))
+
+      saveNewAddresses(addresses, { walletName, mnemonic })
+
+      deriveAddressesInGroupsWorker.terminate()
+    }
+
+    deriveAddressesInGroupsWorker.postMessage({
+      mnemonic: mnemonic,
+      groups,
+      skipIndexes: currentAddressIndexes
+    })
+  }
+
+  const restoreAddressesFromMetadata = async ({
+    mnemonic,
+    walletName,
+    isPassphraseUsed = false
+  }: RestoreAddressesFromMetadataProps) => {
+    const addressesMetadata: AddressMetadata[] = isPassphraseUsed
+      ? []
+      : AddressMetadataStorage.load({ mnemonic, walletName })
+
+    if (addressesMetadata.length > 0) {
+      dispatch(addressRestorationStarted())
+
+      deriveAddressesFromIndexesWorker.onmessage = async ({ data }: { data: AddressKeyPair[] }) => {
+        const restoredAddresses = data.map((address) => {
+          const { isMain, color, ...metadata } = addressesMetadata.find(
+            (metadata) => metadata.index === address.index
+          ) as AddressMetadata
+
+          return {
+            ...address,
+            ...metadata,
+            // TODO: Write a migration script for all addresses with no colors and then remove the following line
+            color: color || getRandomLabelColor(),
+            // TODO: Write a migration script to rename `isMain` to `isDefault` adn then remove the following line
+            isDefault: isMain
+          }
+        })
+
+        dispatch(addressesRestoredFromMetadata(restoredAddresses))
+      }
+
+      deriveAddressesFromIndexesWorker.postMessage({
+        mnemonic: mnemonic,
+        indexesToDerive: addressesMetadata.map((metadata) => metadata.index)
+      })
+    }
+  }
+
+  const discoverAndSaveUsedAddresses = async ({
+    mnemonic,
+    walletName,
+    skipIndexes,
+    enableLoading = true
+  }: DiscoverUsedAddressesProps = {}) => {
+    if (!mnemonic || !walletName) throw new Error('Could not generate addresses, active wallet not found')
+
+    addressDiscoveryWorker.onmessage = ({ data }: { data: AddressKeyPair[] }) => {
+      const addresses: AddressBase[] = data.map((address) => ({
+        ...address,
+        isDefault: false,
+        color: getRandomLabelColor()
+      }))
+
+      saveNewAddresses(addresses, { walletName, mnemonic })
+
+      dispatch(addressDiscoveryFinished(enableLoading))
+    }
+
+    dispatch(addressDiscoveryStarted(enableLoading))
+
+    addressDiscoveryWorker.postMessage({
+      mnemonic,
+      skipIndexes: skipIndexes && skipIndexes.length > 0 ? skipIndexes : currentAddressIndexes,
+      clientUrl: client.explorer.baseUrl
+    })
+  }
+
+  return {
+    generateAddress,
+    generateAndSaveOneAddressPerGroup,
+    restoreAddressesFromMetadata,
+    discoverAndSaveUsedAddresses
+  }
+}
+
+export default useAddressGeneration
```

Comment:
I am curious about your opinion on this whole file. My idea was to group together all "address generation" functions. The main reason is that all these functions are "expensive" cryptographic operations and utilize web workers to make the experience smoother. I went for a hook for the following reasons:
- Keeps the number of arguments to the functions to the minimum necessary
- Can make use of `useCallback` because some of them are used as deps in `useEffects`

Note that some functions still need optional wallet params (mnemonic, walletName) because they can be used in places where the store hasn't yet been informed about the new wallet (wallet creation/import page).

Reply:
The more I think about this, the more I believe that some cleaning of what we call "utils" could help us a lot.
It's even more true now that we have those `storage-utils`. What makes some functions go into the latter folder vs. the generic utils folder?
Why is `loadContacts` in the generic `utils/contacts.ts` file and not in the `storage-utils`, when it's pretty much focused on loading things from our storage?

Should we maybe abstract things a little bit more? Ie. organise things by [feature / topic rather than technicalities & file types](https://reactjs.org/docs/faq-structure.html) (hooks vs. storage vs...)? 

I'm throwing rough thoughts on the table to start a discussion, I don't have a perfect solution in mind just yet. But I really feel like we could squash and merge some files and make our project structure slimmer and less confusing :)

Reply:
A rough first tree...

```
features
├── wallet
│   ├── walletCreation.ts
│   ├── walletSwitch.ts
│   └── ...
├── addresses
│   ├── addressGeneration.ts
│   ├── addressSettingsUpdate.ts
│   └── ...
├── transactions
│   ├── transactionSend.ts
│   ├── transactionDisplay.ts // format outputs, get amounts etc.
│   └── ...
├── theme
│   ├── themeSwitch.ts
│   └── ...
└── amounts
    ├── amountsFormating.ts
    └── ...
...
```

Reply:
> Is it really necessary to separate our hooks from the other utils functions? We know from their names (use prefix) what they are.

I don't see any benefits in having all hooks grouped together. So no, not necessary.

> The more I think about this, the more I believe that some cleaning of what we call "utils" could help us a lot.
It's even more true now that we have those storage-utils. What makes some functions go into the latter folder vs. the generic utils folder?

I agree on reconsidering what we call "utils". As for the question, the functions that focus on app/persistent storage go to storage-utils. I think that's an easy distinction.

> Why is loadContacts in the generic utils/contacts.ts file and not in the storage-utils, when it's pretty much focused on loading things from our storage?

Because storage-utils was introduced after loadContracts was first introduced and I did not remember that I had created it in the past :)

> But I really feel like we could squash and merge some files and make our project structure slimmer and less confusing :)

Yes!

> A rough first tree...

Would that mean we will also put our components, api, and other files inside the features folder? Or is this only about utils? :/

Reply:
I have the feeling that if we go for a features folder structure then everything needs to go into it. Wouldn't it be weird to have function related to wallet creation inside `@/features/wallet/walletGeneration.ts` but the JSX of the screen to be at `@/pages/NewWallet/CreateWalletPage.tsx`?

I wanna put more thought into this, but I also want to keep implementing features of the v2.0 release so that we're not late!

Reply:
One thing that is quite clear to me is that I would like things that are general and are used in many places to be grouped by type (think basic components like buttons, I think they should stay at `@/components/buttons`) but I would definitely like a closer coupling of things that are not general and are very related to each other (for example, I don't like that we have `@/hooks/useGasSettings.ts`, `@/modals/SendModals/SendModalTransfer.tsx`, and `@/modals/SendModals/GasSettingsExpandableSection.tsx`)

In your proposal above I see that you grouped features together, but keep in mind that many might be related to more than 1 category, leading to confusion when looking for something.

I think it'd be easier if we had a call about this, every time I am writing a proposal I have to scratch it because I immediately think of drawbacks.

Reply:
To clarify : my proposal would mean that we (try to) split presentational functions (pages, components, modals) on one side and logic / utils functions on another side. `pages`, `components` and `modals` could be placed into a `ui` parent folder. Those presentational components are making use of the logic placed in the `features` folder. But then, our tree goes deeper, and I don't know if that's something we want. 

Anyway, I wanted to raise the topic now, because I wanted to know if that was something that bothered you in the past. But we can definitely postpone the (potential) changes to later, and schedule a call when we wanna talk more about this.

Reply:
I really like the idea of `features` a lot! I am just not sure whether the presentational stuff should be separate from them. I'll keep working on v2.0 and if I have any good inspiration and ideas I will let you know!